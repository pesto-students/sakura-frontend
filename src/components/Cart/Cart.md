```js
<Cart size={2} sizeUnit="rem" cartSize={30}  cartCbk = {(query) =>{console.log(query)}}></Cart>
```

