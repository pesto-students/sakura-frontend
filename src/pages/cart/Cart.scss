@import "../../sassStyles/globals";
@import "~bootstrap/scss/bootstrap";

.primary_cart{
    width: 100%;
    &_heading{
        margin: 2rem auto;
        font-size: 2rem;
        text-align: center;
        color:$secondary-color;
    }
    &_details{
        margin: 1rem 15vw;

        &_checkout{
            position: -webkit-sticky;
            position: sticky;
            top: 0;
            display: grid;
            width: 100%;
            display: grid;
            box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
            padding: 0.2rem 1rem;
            grid-template-columns: 1fr 1fr 1fr;
            grid-template-rows: repeat(6, min-content);
            border-radius: 4px;

            &_heading{
                grid-column: 1/4;
                font-size: 1.8rem;
                font-weight: 400;
            }
            &_item_amount_heading {
                grid-column: 1/3;
                font-size: 1.3rem;
            }

            &_item_amount_total {
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
            }

            &_shipping_amount_heading{
                grid-column: 1/3; 
                font-size: 1.3rem;
                

            }
            &_shipping_amount_total{
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
                
            }

            &_tax_amount_heading{
                grid-column: 1/3;
                font-size: 1.3rem;

            }
            &_tax_amount_total{
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
                
            }
            &_divider{
                align-self: center;
                grid-column: 1/4;
            }

            &_total_amount_heading{
                grid-column: 1/3;
                font-size: 1.3rem;
                align-self: start;
                padding-bottom: 1rem;
            }

            &_total_amount_total{
                grid-column: 3/4;
                font-size: 1.3rem;
                justify-self: end;
                
                font-weight: 400;
                align-self: start;
                padding-bottom: 1rem;
            }
            &_payment_button{
                grid-column: 1/4;
                padding-bottom: 2rem;
            }
        }

        &_empty{
        font-size: 1.5rem;
        text-align: center;
        color:$secondary-color;
        padding: 1rem;
        }
    }
    
}   

.form-control:focus {
    border-color: $primary-color !important;
    box-shadow: inset 0 1px 1px rgba(0, 0, 0, 0.075), 0 0 8px rgba($primary-color, 0.6) !important;
  }

.form-check-input:focus {
    border-color: $primary-color !important;
    outline: 0;
    box-shadow: 0 0 0 0.25rem rgba($primary-color, 0.2)  !important;
}
.form-check-input:checked {
    background-color:  $primary-color !important;
    border-color:  $primary-color !important;
}

