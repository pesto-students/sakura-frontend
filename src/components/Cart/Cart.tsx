import React from "react";
import { Cart3 } from "react-bootstrap-icons";
import "./Cart.scss";

/**
 * This component tells how many items are there in user cart
 */
export default class Cart extends React.Component<CartProps> {
  constructor(props: CartProps) {
    super(props);
    this.state = {};
  }

  render() {
    const { size = 2, sizeUnit = "rem", cartItemCount: cartItemLength = 0 } = this.props;
    let iconSize: string = String(size * 1.3) + sizeUnit;
    let counterSize: string = String(size) + sizeUnit;

    return (
      <div
        className="primary_cart"
        data-testid="cart"
        onClick={(e) => this.props.cartCbk()}
      >
        <Cart3
          className="primary_cart_icon"
          style={{ width: iconSize, height: iconSize }}
        />
        <div
          className="primary_cart_counter"
          data-testid="cart-counter"
          style={{
            width: counterSize,
            height: counterSize,
            lineHeight: counterSize,
          }}
        >
          {cartItemLength}
        </div>
      </div>
    );
  }
}

export type CartProps = {
  /**
   * Size defines the size of component on different screen.
   */
  size?: number;
  /**
   * Define unit of size
   */
  sizeUnit?: String;
  /**
   * Cart item length tells the number of item in user favorite.
   */
  cartItemCount: number;
  /**
   * Cart callback to be called on click
   */
  cartCbk: () => void;
};
