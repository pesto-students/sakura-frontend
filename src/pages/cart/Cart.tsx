import React, { Component, Fragment, useEffect, useState } from "react";
import { push } from "connected-react-router";
import "./Cart.scss";

import Footer from "../../components/Footer";
import Header from "../header";
import { ErrorBoundary } from "../../components/ErrorBoundary/ErrorBoundary";
import Card2 from "../../components/Card2";
import Test from "../../pages/Test/Test";
import { Col, Row } from "react-bootstrap";
import Button from "../../components/Button";
import { useAppDispatch, useAppSelector } from "../../appStore/hooks";
import {
  getItemFromCart,
  addItemToCart,
  updateItemToCart,
  removeItemToCart,
  emptyCart,
} from "../cart/cart.slice";
import {
  showLoginModal,
  showLogoutModal,
} from "../authenticator/authenticator.slice";
import { addItemToFavorite } from "../favorite/favorite.slice";
import { Modal, Form } from "react-bootstrap";

export default function Cart(props: any) {
  const [amount, setAmount] = useState({
    item: 0,
    tax: 0,
    shipping: 0,
  });
  // modal state and function
  const [showModal, setShowModal] = useState(false);
  const handleModalClose = () => setShowModal(false);
  const handleModalShow = () => setShowModal(true);
  const [showOrderPlacedModal, setShowOrderPlacedModal] = useState(false);
  const handleOrderPlacedModalClose = () => setShowOrderPlacedModal(false);
  const handleOrderPlacedModalShow = () => setShowOrderPlacedModal(true);

  const cartItems = useAppSelector((state) => state.cart.cartItems);
  const userState = useAppSelector((state) => state.auth.userState);
  const dispatch = useAppDispatch();

  useEffect(() => {
    dispatch(getItemFromCart({}));
    // TODO: call api and get items currently present in favorite and cart
    // TODO: useAppSelector to get precise state of products
  }, []);

  useEffect(() => {
    let itemTotal: number = cartItems.reduce(
      (a: any, b: any) =>
        parseInt(a) +
        (parseInt(b["discountedPrice"]) * parseInt(b["quantity"]) || 0),
      0
    );
    setAmount((oldvalue) => ({
      ...oldvalue,
      item: Math.floor(itemTotal),
      shipping: Math.floor(itemTotal / 100),
      tax: Math.floor((itemTotal * 18) / 100),
    }));
  }, [cartItems]);

  const populateCartItems = () => {
    return cartItems.map((item: any, index: number) => {
      let productDescription = {
        brandName: item.brandName,
        productName: item.productName,
        productColor: item.color,
        productSize: item.size,
        originalPrice: `INR ${item.originalPrice}`,
        discountedPrice: `INR ${item.discountedPrice}`,
        productImage: item.productImage,
        productId: item.productId,
        quantity: item.quantity,
      };
      return (
        <div
          style={{
            paddingBottom: "1rem",
            display: "inline-block",
            width: "100%",
          }}
          key={index}
        >
          <Card2
            height="15rem"
            showAddtoFavoriteButton={true}
            productDesc={productDescription}
            handleQuantityChange={(value) => {
              dispatch(updateItemToCart({ ...item, quantity: value }));
            }}
            handleAddToDeleteClick={(productId) => {
              dispatch(removeItemToCart({ productId }));
            }}
            handleAddToFavoriteClick={(productId) => {
              dispatch(removeItemToCart({ productId }));
              dispatch(addItemToFavorite(item));
            }}
          />
        </div>
      );
    });
  };

  const collectInfoModal = () => {
    return (
      <Modal show={showModal} onHide={handleModalClose}>
        <Modal.Header>
          <Modal.Title>Checkout</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Form>
            <Form.Group className="mb-3" controlId="formBasicEmail">
              <Form.Label>Email address</Form.Label>
              <Form.Control
                type="email"
                placeholder="Enter email*"
                value={userState?.userData?.email}
                required={true}
                disabled={true}
              />
              <Form.Text className="text-muted">
                We'll never share your email with anyone else.
              </Form.Text>
            </Form.Group>
            <Form.Group className="mb-3" controlId="formBasicPassword">
              <Form.Label>Phone Number</Form.Label>
              <Form.Control
                type="text"
                placeholder="Enter phone no*"
                required={true}
              />
            </Form.Group>
            <Form.Group className="mb-3" controlId="formBasicPassword">
              <Form.Label>Address</Form.Label>
              <Form.Control
                type="text"
                placeholder="Enter Address*"
                required={true}
              />
            </Form.Group>
            <Form.Group className="mb-3" controlId="formBasicCheckbox">
              <Form.Check
                type="checkbox"
                label="Cash on Delivery"
                checked={true}
              />
            </Form.Group>
            <Button
              width="8rem"
              buttonText="Place Order"
              handleOnClick={(e) => {
                e.preventDefault();
                handleModalClose();
                handleOrderPlacedModalShow();
                dispatch(emptyCart({}));
              }}
            />
            &nbsp; &nbsp;Amount to be paid{" "}
            <b>{`INR ${amount.item + amount.tax + amount.shipping}`}</b>
          </Form>
        </Modal.Body>
        <Modal.Footer>
          <Button
            width="4rem"
            buttonText="Close"
            handleOnClick={handleModalClose}
          />
        </Modal.Footer>
      </Modal>
    );
  };

  const orderPlacedModal = () => {
    return (
      <Modal show={showOrderPlacedModal} onHide={handleOrderPlacedModalClose}>
        <Modal.Header>
          <Modal.Title>Order status</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          Order placed successfully. Order Id:{" "}
          <b>{Date.now() + String(Math.floor(Math.random() * 10000))}</b>
        </Modal.Body>
        <Modal.Footer>
          <Button
            width="4rem"
            buttonText="Close"
            handleOnClick={handleOrderPlacedModalClose}
          />
        </Modal.Footer>
      </Modal>
    );
  };

  return (
    <div className="primary_cart">
      {/* header */}
      {/* page heading */}
      <div className="primary_cart_heading">Shopping Cart</div>
      {/* cart details container */}
      <div className="primary_cart_details">
        {cartItems.length > 0 ? (
          <Row>
            <Col sm={8}>
              <div className="primary_cart_details_items">
                {populateCartItems()}
              </div>
            </Col>
            <Col sm={4}>
              <div className="primary_cart_details_checkout">
                <div className="primary_cart_details_checkout_heading">
                  The total amount of
                </div>
                <div className="primary_cart_details_checkout_item_amount_heading">
                  Item
                </div>
                <div className="primary_cart_details_checkout_item_amount_total">
                  {`INR ${amount.item}`}
                </div>
                <div className="primary_cart_details_checkout_shipping_amount_heading">
                  Shipping(10%)
                </div>
                <div className="primary_cart_details_checkout_shipping_amount_total">
                  {`INR ${amount.shipping}`}
                </div>
                <div className="primary_cart_details_checkout_tax_amount_heading">
                  Tax(18%)
                </div>
                <div className="primary_cart_details_checkout_tax_amount_total">
                  {`INR ${amount.tax}`}
                </div>
                <hr className="primary_cart_details_checkout_divider" />
                <div className="primary_cart_details_checkout_total_amount_heading">
                  Total amount
                </div>
                <div className="primary_cart_details_checkout_total_amount_total">
                  {`INR ${amount.item + amount.tax + amount.shipping}`}
                </div>
                <div className="primary_cart_details_checkout_payment_button">
                  <Button
                    width="100%"
                    buttonText="Checkout"
                    handleOnClick={(e) => {
                      userState.isLoggedIn?handleModalShow():dispatch(showLoginModal())
                    }}
                  />
                </div>
              </div>
            </Col>
          </Row>
        ) : (
          <Fragment>
            <div className="primary_cart_details_empty">
              Your Cart is empty. &nbsp;&nbsp;
            </div>{" "}
            <div style={{ textAlign: "center" }}>
              <Button
                width="30rem"
                buttonText="Click here to Explore Products"
                handleOnClick={(e) => {
                  dispatch(push(`/`));
                }}
              />
            </div>{" "}
          </Fragment>
        )}
      </div>

      {/* Footer */}
      <Footer />
      {collectInfoModal()}
      {orderPlacedModal()}
    </div>
  );
}
