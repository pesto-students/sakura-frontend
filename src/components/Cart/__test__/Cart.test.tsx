import { render, fireEvent, waitFor, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from "react-test-renderer"
import Cart from "../Cart"

// const server = setupServer(
//   rest.get('/search', (req, res, ctx) => {
//     return res(ctx.json({}))
//   })
// )

// beforeAll(() => server.listen())
// afterEach(() => server.resetHandlers())
// afterAll(() => server.close())

test('Should  call Cart callback on click', async () => {
  
  render(<Cart size={2} sizeUnit="rem" cartItemCount={10} cartCbk={cartCllback}/>)
  let st = ""

  function cartCllback() {
        st = "cart clicked! go to cart page"
        expect(st).toEqual("cart clicked! go to cart page")
  }

  const cartClick = screen.getByTestId("cart");
  fireEvent.click(cartClick);

})


test('Should render cartItemLegth', async () => {
  
  const cartProps = {size:2, sizeUnit:"rem", cartItemLength: 20, cartCbk: function () {}}
  render(<Cart size={cartProps.size} sizeUnit={cartProps.sizeUnit} cartItemCount={cartProps.cartItemLength} cartCbk={cartProps.cartCbk}/>)
  const cartElement = screen.getByTestId("cart");
  const cartCounterElement = screen.getByTestId("cart-counter");
  expect(cartElement).toBeInTheDocument()
  expect(cartCounterElement).toHaveTextContent("20")

})

test("matches snapshots", ()=> {
  const cartProps = {size:2, sizeUnit:"rem", cartItemLength: 20, cartCbk: function () {}}
  const tree = renderer.create(<Cart size={cartProps.size} sizeUnit={cartProps.sizeUnit}  cartItemCount={cartProps.cartItemLength} cartCbk={cartProps.cartCbk}/>).toJSON()
  expect(tree).toMatchSnapshot()

})