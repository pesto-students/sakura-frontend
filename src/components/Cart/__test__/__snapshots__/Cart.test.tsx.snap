// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshots 1`] = `
<div
  className="primary_cart"
  data-testid="cart"
  onClick={[Function]}
>
  <svg
    className="primary_cart_icon"
    fill="currentColor"
    height="1em"
    style={
      Object {
        "height": "2.6rem",
        "width": "2.6rem",
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 1.5A.5.5 0 0 1 .5 1H2a.5.5 0 0 1 .485.379L2.89 3H14.5a.5.5 0 0 1 .49.598l-1 5a.5.5 0 0 1-.465.401l-9.397.472L4.415 11H13a.5.5 0 0 1 0 1H4a.5.5 0 0 1-.491-.408L2.01 3.607 1.61 2H.5a.5.5 0 0 1-.5-.5zM3.102 4l.84 4.479 9.144-.459L13.89 4H3.102zM5 12a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm7 0a2 2 0 1 0 0 4 2 2 0 0 0 0-4zm-7 1a1 1 0 1 1 0 2 1 1 0 0 1 0-2zm7 0a1 1 0 1 1 0 2 1 1 0 0 1 0-2z"
    />
  </svg>
  <div
    className="primary_cart_counter"
    data-testid="cart-counter"
    style={
      Object {
        "height": "2rem",
        "lineHeight": "2rem",
        "width": "2rem",
      }
    }
  >
    20
  </div>
</div>
`;
